KAFKA-5342: Clarify fatal/abortable exceptions used in producer

hachikuji opened the pull request:
This patch improves documentation on the handling of errors for the idempotent/transactional producer. It also fixes a couple minor inconsistencies and improves test coverage. In particular:
- UnsupportedForMessageFormat should be a fatal error for TxnOffsetCommit responses
- UnsupportedVersion should be fatal for Produce responses and should be returned instead of InvalidRequest

hachikuji committed:
KAFKA-5342: Clarify fatal/abortable exceptions used in producer

hachikuji commented:
cc @apurvam @ijuma

hachikuji committed:
Fix unused import

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -515,34 +515,42 @@ private static int parseAcks(String acksString) {
+     *         does not support transactions (i.e. if its version is lower than 0.11.0.0)
+     * @throws org.apache.kafka.common.errors.AuthorizationException fatal error indicating that the configured
+     *         transactional.id is not authorized
+     * @throws KafkaException if the producer has encountered a previous fatal error or for any other unexpected error
```
Here and in the rest of the methods, the `KafkaException` is not fatal right. The producer can actually abort the transaction in this case? Would be worth adding that to the javadoc.

hachikuji replied:
Ack. I'll attempt to clarify.

apurvam reviewed (commented):
Looks good overall, left one minor comment clarifying the semantics of the `KafkaException`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -688,6 +722,7 @@ public void abortTransaction() throws ProducerFencedException {
      * @param callback A user-supplied callback to execute when the record has been acknowledged by the server (null
      *        indicates no callback)
      *
+     * @throws IllegalStateException if a transactional.id has been configured and no transaction has been started
```
Hmm, we mention in other methods two possible reasons why `UnsupportedVersionException` can be thrown:

1. Broker is too old.
2. Offsets topic is configured to use an older message format.

What if the topic you are sending to is configured with an older message format?

hachikuji replied:
This is also possible. I will add a note to the `send` javadoc.

ijuma replied:
Seems like this wasn't added, or did I miss something?

apurvam replied:
It seems to be added on line 703.

apurvam replied:
Oh you mean `UnsupportedForMessageFormatException`? That doesn't seem to be added.

ijuma replied:
I was referring to the @throws clauses at the bottom where neither seems to have been added. Seems like both are possible.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -274,11 +288,24 @@ public void prepareResponseFrom(AbstractResponse response, Node node, boolean di
+
+    /**
+     * Raise an unsupported version error on the next request. Raise an exception if the next request
+     * does not match the provided matcher function.
```
Hmm, does this throw an unsupported version error if the next request matches the matcher and a different exception if it does not? The description could be a bit clearer.

ijuma reviewed (commented):
Thanks for the PR. Looks good overall, just a couple of minor questions/comments.

hachikuji committed:
Fix a few error code inconsistencies

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1041,6 +1041,9 @@ public void handleResponse(AbstractResponse response) {
                     return;
+                } else if (error == Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT) {
+                    fatalError(error.exception());
+                    return;
```
Should we be using a single `else if` for all the fatal errors?

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/UnsupportedVersionException.java:
```diff
@@ -16,6 +16,11 @@
 
+/**
+ * Indicates that a request API or version needed by the client is not supported by the broker. This is
+ * typically a fatal error as Kafka clients will downgrade request versions as needed except in cases where
```
How do we define `fatal`?

ijuma replied:
To elaborate a little more, if you use a consumer method that returns `UnsupportedVersionException`, you may be able to fallback to an alternative (say `offsetForTimes`). Do we consider that `fatal`?

hachikuji committed:
A couple more clarifications

apurvam commented on clients/src/main/java/org/apache/kafka/common/errors/UnsupportedForMessageFormatException.java:
```diff
@@ -17,7 +17,8 @@
 /**
- * The message format version does not support the requested function.
+ * The message format version does not support the requested function. For example, if idempotence is
+ * requested and the broker is on an old message format, then this error will be returned
```
nit: " and the broker is on a message format older than 0.11.0"

apurvam reviewed (approved):
LGTM! This is a good improvement. I think we should file a JIRA to discuss the idea of a 'FatalKafkaException' class so that this can be codified. If any method returns an instance of `FatalKafkaException`, then the only option for the application is to close the producer. 

We can then workout when to raise fatal exceptions. I think introducing this type would greatly simplify application logic.

hachikuji committed:
Cleanup in UnsupportedForMessageFormat javadoc

ijuma commented:
An alternative to a `FatalKafkaException` class is introducing a `isFatal` method to `KafkaException`. That makes a bit more dynamic.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -492,11 +494,77 @@ public boolean matches(AbstractRequest body) {
+        prepareAndReceiveInitProducerId(producerId, Errors.NONE);
+        assertTrue(transactionManager.hasProducerId());
+
+        // cluster authorization is a fatal error for the producer
```
Is this comment relevant here?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -492,11 +494,77 @@ public boolean matches(AbstractRequest body) {
+        assertTrue(future.isDone());
+        try {
+            future.get();
+            fail("Future should have raised ClusterAuthorizationException");
```
Does this need to be updated?

ijuma reviewed (commented):
A few more minor comments. Thanks for adding more tests and the related fixes. We should probably update the PR and JIRA description to mention that we're actually fixing bugs too in this PR.

hachikuji committed:
Address pending comments

hachikuji committed:
Missed a few comments

hachikuji commented:
@ijuma Thanks for the review. I updated the description and I think I've addressed all comments. I'll update the JIRA as soon as I can access it again.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -715,6 +715,12 @@ public void abortTransaction() throws ProducerFencedException {
      * producer and create a new instance.
      * </p>
      * <p>
+     * If the message format of the destination topic is not upgraded to 0.11.0.0, produce requests will
```
Do we need to make it clear that this is only the case if idempotence=true? And do we want to add additional @throws lines?

hachikuji replied:
Sure I can clarify. The awkwardness of adding `@throws` lines is that `send` itself does not throw the exception: it is raised in the callback.

ijuma replied:
You're right, we should not add the @throws lines then.

ijuma reviewed (approved):
LGTM, just one minor comment.

hachikuji committed:
Emphasize that UnsupportedForMessageFormat applies for idempotent/transactional requests

hachikuji commented:
Merging to trunk and 0.11.0.
